pkg/kubelet/kubelet_test.go

```diff
@@ -2154,6 +2156,65 @@ func TestMakeEnvironmentVariables(t *testing.T) {
 				"KUBERNETES_RO_PORT_8087_TCP_ADDR=1.2.3.7"),
 			21,
 		},
+		{
+			"simple substitution",
+			&api.Pod{ObjectMeta: api.ObjectMeta{Namespace: "unknown", Name: "test-pod"}},
+			&api.Container{
+				Env: []api.EnvVar{
+					{Name: "BAR", Value: "bar"},
+					{Name: "FOO", Value: "${BAR}"},
+				},
+			},
+			"kubernetes",
+			true,
+			util.NewStringSet(
+				"BAR=bar",
+				"FOO=bar",
+			),
+			2,
+		},
+		{
+			"slightly more complex substitution",
+			&api.Pod{ObjectMeta: api.ObjectMeta{Namespace: "unknown", Name: "test-pod"}},
+			&api.Container{
+				Env: []api.EnvVar{
+					{Name: "BAR", Value: "bar"},
```

Comment:
Yes, maybe they should.

What I think is great about all of the approaches discussed here is that the substitution is performed by the control plane, which preserves separation of dev/ops concerns.

By contrast, a downward API that would be directed read/polled/watched by applications would bake that API into many more thousands of applications than those otherwise targeting the management API. Consequently, such an API would need to be narrower and more slowly changing than the management API, since it would be much harder to change.